readthedocs/projects/forms.py

```diff
@@ -91,14 +90,105 @@ def save(self, commit=True):
 
 
 class ProjectBackendForm(forms.Form):
-
     """Get the import backend."""
 
     backend = forms.CharField()
 
 
-class ProjectFormPrevalidateMixin:
+class ProjectPRBuildsMixin(PrevalidatedForm):
+    """
+    Mixin that provides a method to setup the external builds option.
+
+    TODO: Remove this once we migrate to the new dashboard,
+    and we don't need to support the old project settings form.
+    """
+
+    def has_supported_integration(self, integrations):
+        supported_types = {Integration.GITHUB_WEBHOOK, Integration.GITLAB_WEBHOOK}
+        for integration in integrations:
+            if integration.integration_type in supported_types:
+                return True
+        return False
+
+    def can_build_external_versions(self, integrations):
+        """
+        Check if external versions can be enabled for this project.
+
+        A project can build external versions if:
+
+        - They are using GitHub or GitLab.
+        - The repository's webhook is setup to send pull request events.
+
+        If the integration's provider data isn't set,
+        it could mean that the user created the integration manually,
+        and doesn't have an account connected.
+        So we don't know for sure if the webhook is sending pull request events.
+        """
+        for integration in integrations:
+            provider_data = integration.provider_data
+            if integration.integration_type == Integration.GITHUB_WEBHOOK and (
+                not provider_data or "pull_request" in provider_data.get("events", [])
+            ):
+                return True
+            if integration.integration_type == Integration.GITLAB_WEBHOOK and (
+                not provider_data or provider_data.get("merge_requests_events")
+            ):
+                return True
+        return False
 
+    def setup_external_builds_option(self):
+        """Disable the external builds option if the project doesn't meet the requirements."""
+        if (
+            settings.ALLOW_PRIVATE_REPOS
+            and self.instance.remote_repository
+            and not self.instance.remote_repository.private
+        ):
+            self.fields["external_builds_privacy_level"].disabled = True
+            help_text = _(
+                "We have detected that this project is public, pull request previews are set to public."
+            )
```

Comment:
```suggestion
            # TODO use a proper error/warning instead of help text for error states
            help_text = _(
                "We have detected that this project is public, pull request previews are set to public."
            )
```